src/v/cluster/members_backend.cc

```diff
@@ -103,79 +128,257 @@ void members_backend::start_reconciliation_loop() {
+    return o;
+};
+
+void members_backend::calculate_reallocations_after_node_added(
+  members_backend::update_meta& meta) {
+    auto& topics = _topics.local().topics_map();
+    struct node_info {
+        size_t replicas_count;
+        size_t max_capacity;
+    };
+    // 1. count current node allocations
+    absl::node_hash_map<model::node_id, node_info> node_replicas;
+
+    size_t total_replicas = 0;
+    for (const auto& [id, n] : _allocator.local().state().allocation_nodes()) {
+        if (!node_replicas.contains(id)) {
+            node_replicas.emplace(
+              id,
+              node_info{
+                .replicas_count = 0,
+                .max_capacity = n->max_capacity(),
+              });
+        }
+        auto it = node_replicas.find(id);
+        it->second.replicas_count += n->allocated_partitions();
+        total_replicas += n->allocated_partitions();
+    }
+
+    // 2. calculate number of replicas per node leading to even replica per node
+    // distribution
+    auto target_replicas_per_node = total_replicas
+                                    / _members.local().all_brokers().size();
+
+    // 3. calculate how many replicas have to be moved from each node
+    std::vector<replicas_to_move> to_move_from_node;
+    for (auto& [id, info] : node_replicas) {
+        auto to_move = std::max<size_t>(
+          0, info.replicas_count - target_replicas_per_node);
+        to_move_from_node.emplace_back(id, to_move);
+    }
+
+    vlog(
+      clusterlog.info,
+      "Targeting to move {} partitions and reallocate them to node {}",
+      to_move_from_node,
+      meta.update.id);
+
+    auto cmp = [](const replicas_to_move& lhs, const replicas_to_move& rhs) {
+        return lhs.left_to_move < rhs.left_to_move;
+    };
+
+    // 4. Pass over all partition metadata once, try to move until we reach even
+    // number of partitons per core on each node
+    for (auto& [tp_ns, metadata] : topics) {
+        // do not try to move internal partitions
+        if (
+          tp_ns.ns == model::kafka_internal_namespace
+          || tp_ns.ns == model::redpanda_ns) {
+            continue;
+        }
+        for (auto& p : metadata.assignments) {
+            std::erase_if(to_move_from_node, [](const replicas_to_move& v) {
+                return v.left_to_move == 0;
+            });
+
+            std::sort(to_move_from_node.begin(), to_move_from_node.end(), cmp);
+            for (auto& to_move : to_move_from_node) {
+                if (is_in_replica_set(p.replicas, to_move.id)) {
+                    partition_reallocation reallocation(
+                      model::ntp(tp_ns.ns, tp_ns.tp, p.id), p.replicas.size());
+
+                    reallocation.replicas_to_remove.emplace(to_move.id);
+                    meta.partition_reallocations.push_back(
+                      std::move(reallocation));
+                    to_move.left_to_move--;
```

Comment:
`left_to_move` is unsigned and can underflow, maybe the check should be added that verifies that `left_to_move` is not 0, or comment that explains why it will never underflow (probably, there is some loop invariant that prevents this value from underflow that I'm missing, it looks like `erase_if` online 241 assumes that this value goes down to 0)